samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -0,0 +1,172 @@
+   * <p>
+   * <b>Note:</b>This identifier has to be unique across the instances of StreamProcessors.
+   */
+  private static final String PROCESSOR_ID = "processor.id";
+  private final int processorId;
+  private final JobCoordinator jobCoordinator;
+  private final SamzaContainerController containerController;
+
+  /**
+   * Create an instance of StreamProcessor that encapsulates a JobCoordinator and Samza Container
+   * <p>
+   * JobCoordinator controls how the various StreamProcessor instances belonging to a job coordinate. It is also
+   * responsible generating and updating JobModel.
+   * When StreamProcessor starts, it starts the JobCoordinator and brings up a SamzaContainer based on the JobModel.
+   * SamzaContainer is executed using an ExecutorService.
+   * <p>
+   * <b>Note:</b> Lifecycle of the ExecutorService is fully managed by the StreamProcessor, and NOT exposed to the user
+   *
+   * @param processorId            Unique identifier for a processor within the job. It has the same semantics as
+   *                               "containerId" in Samza
+   * @param config                 Instance of config object - contains all configuration required for processing
+   * @param customMetricsReporters Map of custom MetricReporter instances that are to be injected in the Samza job
+   * @param asyncStreamTaskFactory The {@link AsyncStreamTaskFactory} to be used for creating task instances.
+   */
+  public StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                         AsyncStreamTaskFactory asyncStreamTaskFactory) {
+    this(processorId, config, customMetricsReporters, (Object) asyncStreamTaskFactory);
+  }
+
+  /**
+   * Same as {@link #StreamProcessor(int, Config, Map, AsyncStreamTaskFactory)}, except task instances are created
+   * using the provided {@link StreamTaskFactory}.
+   */
+  public StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                         StreamTaskFactory streamTaskFactory) {
+    this(processorId, config, customMetricsReporters, (Object) streamTaskFactory);
+  }
+
+  /**
+   * Same as {@link #StreamProcessor(int, Config, Map, AsyncStreamTaskFactory)}, except task instances are created
+   * using the "task.class" configuration instead of a task factory.
+   */
+  public StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters) {
+    this(processorId, config, customMetricsReporters, (Object) null);
+  }
+
+  private StreamProcessor(int processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                          Object taskFactory) {
+    this.processorId = processorId;
+
+    Map<String, String> updatedConfigMap = new HashMap<>();
+    updatedConfigMap.putAll(config);
+    updatedConfigMap.put(PROCESSOR_ID, String.valueOf(this.processorId));
```

Comment:
Thanks a lot for the details. It's much clear to me now after seeing the two grouper factory classes. Do we want to make the config public (something like adding to the container config)? I am not sure whether there is any use case that customer uses standalone with their own groupers, and they need to access the processId there too.